File: app/src/main/java/com/mstc/mstcapp/MainActivity.java
package com.mstc.mstcapp;

import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;

import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.FragmentManager;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.google.android.material.chip.Chip;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.mstc.mstcapp.ui.ProjectIdeaFragment;
import com.mstc.mstcapp.util.Constants;

import java.util.HashMap;
import java.util.Random;

public class MainActivity extends AppCompatActivity {
    public static final HashMap<String, Boolean> fetchedData = new HashMap<>();
    public static boolean isAppRunning = false;
    public static int feed_position = 0;
    public static boolean isHome = false;
    private final Context context = this;
    private final int[] ids = {R.id.home, R.id.resources, R.id.explore};
    int exitCount = 0;
    private NavController navController;
    private DrawerLayout drawerLayout;

    public static int getFeed_position() {
        return feed_position;
    }

    public static void setFeed_position(int feed_position) {
        MainActivity.feed_position = feed_position;
    }

    public static boolean isFetchedData(String domain) {
        if (fetchedData.containsKey(domain))
            return fetchedData.get(domain);
        else
            return false;
    }

    public static void setFetchedData(String domain) {
        fetchedData.put(domain, true);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        navController = Navigation.findNavController(this, R.id.nav_host_fragment);
        drawerLayout = findViewById(R.id.drawerLayout);

        drawerLayout.findViewById(R.id.share).setOnClickListener(v -> share());
        drawerLayout.findViewById(R.id.feedback).setOnClickListener(v -> openURL("market://details?id=" + context.getPackageName()));
        drawerLayout.findViewById(R.id.idea).setOnClickListener(v -> openIdeaDialog());

        drawerLayout.findViewById(R.id.instagram).setOnClickListener(v -> openURL(Constants.INSTAGRAM_URL));
        drawerLayout.findViewById(R.id.facebook).setOnClickListener(v -> openURL(Constants.FACEBOOK_URL));
        drawerLayout.findViewById(R.id.linkedin).setOnClickListener(v -> openURL(Constants.LINKEDIN_URL));
        drawerLayout.findViewById(R.id.github).setOnClickListener(v -> openURL(Constants.GITHUB_URL));

        drawerLayout.findViewById(R.id.privacy_policy).setOnClickListener(v -> openURL(Constants.PRIVACY_URL));

        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawerLayout.setDrawerListener(toggle);

        toggle.syncState();
        toolbar.setNavigationIcon(ContextCompat.getDrawable(context, R.drawable.ic_navigation));

        findViewById(R.id.home).setOnClickListener(v -> {
            if (isHome) feed_position = 0;
            selectTab(ids[0]);
            navController.popBackStack();
            navController.navigate(R.id.navigation_home);
        });

        findViewById(R.id.resources).setOnClickListener(v -> {
            selectTab(ids[1]);
            navController.popBackStack();
            navController.navigate(R.id.navigation_resources);
        });

        findViewById(R.id.explore).setOnClickListener(v -> {
            selectTab(ids[2]);
            navController.popBackStack();
            navController.navigate(R.id.navigation_explore);
        });
    }

    private void share() {
        Intent intent = new Intent(android.content.Intent.ACTION_SEND);
        Random rand = new Random();

        String[] messages = {"Are you someone who still cannot find the material to start your journey to become a proficient developer?\n",
                "On the lookout for study material?\n",
                "Budding developers! Still on the lookout for study material?\n"
        };
        String message = messages[rand.nextInt(3)] +
                "\nDon’t worry STC has got you covered!\n" +
                "\nDownload the STC app for latest resources and guidelines from \n" + Constants.PLAY_STORE_URL +
                "\n\nGuess what, it is FREE!";
        intent.setType("text/plain");
        intent.putExtra(android.content.Intent.EXTRA_TEXT, message);
        startActivity(Intent.createChooser(intent, "Share Using"));
    }


    private void openIdeaDialog() {
        FragmentManager fm = getSupportFragmentManager();
        ProjectIdeaFragment projectIdeaFragment = ProjectIdeaFragment.newInstance();
        projectIdeaFragment.show(fm, "projectFragment");
    }

    public void openURL(String url) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(url));
        startActivity(intent);
    }

    @Override
    public void onBackPressed() {
        if (drawerLayout.isOpen()) {
            drawerLayout.close();
        } else if (navController.getCurrentDestination().getId() != R.id.navigation_home) {
            selectTab(ids[0]);
            navController.popBackStack();
            navController.navigate(R.id.navigation_home);
        } else {
            exitCount++;
            View view = findViewById(android.R.id.content);
            Snackbar snackbar = Snackbar.make(context, view, "Press back again to exit",
                    BaseTransientBottomBar.LENGTH_SHORT);
            snackbar.addCallback(new Snackbar.Callback() {
                @Override
                public void onDismissed(Snackbar transientBottomBar, int event) {
                    exitCount = 0;
                }
            });
            snackbar.show();
            if (exitCount == 2) {
                super.onBackPressed();
            }
        }
    }

    private void selectTab(int id) {
        for (int j : ids) {
            if (j != id) setUnselected(j);
        }
        setSelected(id);
    }

    private void setUnselected(int id) {
        Chip chip = findViewById(id);
        chip.setChipBackgroundColor(ColorStateList.valueOf(ContextCompat.getColor(context, R.color.white)));
        chip.setTextColor(ContextCompat.getColor(context, R.color.textColorPrimary));
        chip.setChipIconTint(ColorStateList.valueOf(ContextCompat.getColor(context, R.color.textColorPrimary)));
    }

    private void setSelected(int id) {
        Chip chip = findViewById(id);
        chip.setChipBackgroundColor(ColorStateList.valueOf(ContextCompat.getColor(context, R.color.colorTertiaryBlue)));
        chip.setTextColor(ContextCompat.getColor(context, R.color.colorPrimary));
        chip.setChipIconTint(ColorStateList.valueOf(ContextCompat.getColor(context, R.color.colorPrimary)));
    }
}

File: app/src/main/java/com/mstc/mstcapp/MainActivity.kt
package com.mstc.mstcapp

import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.os.Bundle
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.navigation.NavController
import androidx.navigation.Navigation
import com.google.android.material.chip.Chip
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.mstc.mstcapp.databinding.ActivityMainBinding
import com.mstc.mstcapp.util.Constants
import com.mstc.mstcapp.util.Functions.Companion.openURL
import java.util.*


private const val TAG = "MainActivity"

class MainActivity : AppCompatActivity() {
    lateinit var binding: ActivityMainBinding
    val context: Context = this
    private lateinit var navController: NavController

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.toolbar)
        navController = Navigation.findNavController(this, R.id.nav_host_fragment)
        initDrawer()
        binding.apply {
            selectTab(bottomNav.home)
            bottomNav.home.setOnClickListener {
                selectTab(bottomNav.home)
                if (navController.currentDestination?.id != R.id.navigation_home)
                    navController.popBackStack().also {
                        toolbar.title = "Home"
                    }
            }
            bottomNav.resources.setOnClickListener {
                selectTab(bottomNav.resources)
                navController.apply {
                    if (currentDestination?.id != R.id.navigation_home)
                        popBackStack()
                    navigate(R.id.navigation_resources).also {
                        toolbar.title = "Resources"
                    }
                }
            }

            bottomNav.explore.setOnClickListener {
                selectTab(bottomNav.explore)
                navController.apply {
                    if (currentDestination?.id != R.id.navigation_home)
                        popBackStack()
                    navigate(R.id.navigation_explore).also {
                        toolbar.title = "Explore"
                    }
                }
            }

            val toggle = ActionBarDrawerToggle(
                this@MainActivity,
                drawerLayout,
                toolbar,
                R.string.navigation_drawer_open,
                R.string.navigation_drawer_close
            )
            drawerLayout.addDrawerListener(toggle)
            toggle.syncState()
            toolbar.navigationIcon = ContextCompat.getDrawable(
                context,
                R.drawable.ic_navigation
            )
        }
    }


    var exitCount: Int = 0
    override fun onBackPressed() {
        if (navController.currentDestination?.id == R.id.navigation_home) {
            exitCount++
            Snackbar.make(
                context, findViewById(android.R.id.content), "Press back again to exit",
                BaseTransientBottomBar.LENGTH_SHORT
            ).addCallback(object : Snackbar.Callback() {
                override fun onDismissed(transientBottomBar: Snackbar, event: Int) {
                    exitCount = 0
                }
            }).show()
            if (exitCount == 2) {
                super.onBackPressed()
            }
        } else {
            selectTab(binding.bottomNav.home)
            navController.popBackStack()
        }
    }

    private fun initDrawer() {
        binding.navigationDrawer.apply {
            share.setOnClickListener { share() }
            feedback.setOnClickListener { feedback() }
            idea.setOnClickListener {
                navController.navigate(R.id.projectIdeaDialog)
            }
            instagram.setOnClickListener { openURL(context, Constants.INSTAGRAM_URL) }
            facebook.setOnClickListener { openURL(context, Constants.FACEBOOK_URL) }
            linkedin.setOnClickListener { openURL(context, Constants.LINKEDIN_URL) }
            github.setOnClickListener { openURL(context, Constants.GITHUB_URL) }
            privacyPolicy.setOnClickListener { openURL(context, Constants.PRIVACY_URL) }
        }
    }

    private fun feedback() {
        MaterialAlertDialogBuilder(this)
            .setTitle("Give us a feedback")
            .setMessage("This will redirect you to Google Play Store")
            .setPositiveButton("Rate App") { _, _ ->
                openURL(
                    context,
                    "market://details?id=" + context.packageName
                )
            }
            .setNegativeButton("Not Now") { _, _ -> }
            .show()
    }

    private fun selectTab(chip: Chip) {
        binding.bottomNav.apply {
            for (j in arrayOf(home, resources, explore)) {
                setUnselected(j)
            }
            setSelected(chip)
        }
    }

    private fun setUnselected(chip: Chip) {
        chip.chipBackgroundColor =
            ColorStateList.valueOf(ContextCompat.getColor(context, R.color.white))
        ColorStateList.valueOf(ContextCompat.getColor(context, R.color.textColorPrimary))
            .also {
                chip.chipIconTint = it
                chip.setTextColor(it)
            }
    }

    private fun setSelected(chip: Chip) {
        chip.chipBackgroundColor =
            ColorStateList.valueOf(ContextCompat.getColor(context, R.color.colorTertiaryBlue))
        ColorStateList.valueOf(ContextCompat.getColor(context, R.color.colorPrimary))
            .also {
                chip.setTextColor(it)
                chip.chipIconTint = it
            }
    }

    private fun share() {
        val rand = Random()
        val messages =
            arrayOf(
                "Are you someone who still cannot find the material to start your journey to become a proficient developer?\n",
                "On the lookout for study material?\n",
                "Budding developers! Still on the lookout for study material?\n"
            )
        val message = """
            ${messages[rand.nextInt(3)]}
            Don’t worry STC has got you covered!
            
            Download the STC app for latest resources and guidelines from 
            ${Constants.PLAY_STORE_URL}
            
            Guess what, it is FREE!
            """.trimIndent()

        val intent = Intent(Intent.ACTION_SEND)
            .also {
                it.type = "text/plain"
                it.putExtra(Intent.EXTRA_TEXT, message)
            }
        startActivity(Intent.createChooser(intent, "Share Using"))
    }

    companion object {
        private val fetchedData = HashMap<String, Boolean>()
        fun isFetchedData(domain: String): Boolean {
            return if (fetchedData.containsKey(domain))
                fetchedData[domain] == true
            else false
        }

        fun setFetchedData(domain: String) {
            fetchedData[domain] = true
        }
    }
}
